clientd3d/audio_openal.c

```diff
@@ -0,0 +1,843 @@
+// Meridian 59, Copyright 1994-2025 Andrew Kirmse and Chris Kirmse.
+// All rights reserved.
+//
+// This software is distributed under a license that is described in
+// the LICENSE file that accompanies it.
+//
+// Meridian is a registered trademark.
+
+/*
+ * audio_openal.c: OpenAL Soft audio engine implementation for Meridian 59
+ */
+
+#include "client.h"
+#include <AL/al.h>
+#include <AL/alc.h>
+#include <AL/alext.h>
+#include <AL/efx.h>
+
+#include <functional>
+#include <list>
+#include <unordered_map>
+
+// STB Vorbis for OGG decoding (compiled separately with STB_VORBIS_NO_PUSHDATA_API)
+extern "C" int stb_vorbis_decode_filename(const char *filename, int *channels, int *sample_rate, short **output);
+
+// Configuration
+static const int MAX_AUDIO_SOURCES = 32;
+static const int MAX_CACHED_BUFFERS = 256;
+
+// Global state
+static ALCdevice* g_device = NULL;
+static ALCcontext* g_context = NULL;
+static ALuint g_sources[MAX_AUDIO_SOURCES];
+static int g_numSources = 0;
+static bool g_initialized = false;
+
+// LRU buffer cache: list ordered by recency (front = newest), map for O(1) lookup.
+struct CacheNode {
+   std::string filename;
+   ALuint buffer;
+};
+
+struct CaseInsensitiveHash {
+   size_t operator()(const std::string& s) const {
+      std::string lower;
+      lower.reserve(s.size());
```

Comment:
You really don't want to be allocating memory during the computation of a hash function; that more than eliminates all the performance gains of using a hash table vs. a vector.  You should be able to find something online, perhaps using c_str() to avoid the allocation.